src/client/debugger/jupyter/debuggingManager.ts

```diff
@@ -0,0 +1,228 @@
+                    return;
+                }
+            }),
+
+            vscode.commands.registerCommand('jupyter.debugNotebook', () => {
+                const editor = vscode.window.activeNotebookEditor;
+                if (editor) {
+                    this.updateToolbar(true);
+                    void this.startDebugging(editor.document);
+                } else {
+                    void vscode.window.showErrorMessage('No active notebook document to debug');
+                }
+            })
+        );
+    }
+
+    private fixBreakpoints(doc: vscode.NotebookDocument) {
+        const map = new Map<string, vscode.Uri>();
+
+        doc.getCells().forEach((cell, ix) => {
+            const pos = parseInt(cell.document.uri.fragment);
+            if (pos !== ix) {
+                map.set(
+                    cell.document.uri.toString(),
+                    cell.document.uri.with({ fragment: ix.toString().padStart(8, '0') })
+                );
+            }
+        });
+
+        if (map.size > 0) {
+            const addBpts: vscode.SourceBreakpoint[] = [];
+            const removeBpt: vscode.SourceBreakpoint[] = [];
+            for (const b of vscode.debug.breakpoints) {
+                if (b instanceof vscode.SourceBreakpoint) {
+                    const s = map.get(b.location.uri.toString());
+                    if (s) {
+                        removeBpt.push(b);
+                        const loc = new vscode.Location(s, b.location.range);
+                        addBpts.push(
+                            new vscode.SourceBreakpoint(loc /*, b.enabled, b.condition, b.hitCondition, b.logMessage*/)
+                        );
+                    }
+                }
+            }
+            if (removeBpt.length > 0) {
+                vscode.debug.removeBreakpoints(removeBpt);
+            }
+            if (addBpts.length > 0) {
+                vscode.debug.addBreakpoints(addBpts);
+            }
+        }
+    }
+
+    private updateToolbar(debugging: boolean) {
+        this.debuggingInProgress.set(debugging).ignoreErrors();
+    }
+
+    private async startDebugging(doc: vscode.NotebookDocument) {
+        let dbg = this.notebookToDebugger.get(doc);
+        if (!dbg) {
+            dbg = new Debugger(doc);
+            this.notebookToDebugger.set(doc, dbg);
+
+            try {
+                await dbg.session;
+
+                // toggle the breakpoint margin
+                void vscode.commands.executeCommand('notebook.toggleBreakpointMargin', doc);
+            } catch (err) {
+                void vscode.window.showErrorMessage(`Can't start debugging (${err})`);
+            }
+        }
+    }
+
+    private getDebuggerByUri(document: vscode.NotebookDocument): Debugger | undefined {
+        for (const [doc, dbg] of this.notebookToDebugger.entries()) {
+            if (document.uri.toString() === doc.uri.toString()) {
```

Comment:
Per-discussion here: https://github.com/microsoft/vscode-jupyter/pull/6358#discussion_r657594246 I believe we can just compare on documents.

Reply:
cleanup commit